Development

TarasTsalko opened the pull request.

Commit:
Реализуйте два constexpr-конструктора класса Book и добавлен объект
std::formatter для класса Book

Commit:
Добавлены компораторы для элементов класса Book

Commit:
Реализован класс BookDatabase и изменен порядок параметров
в конструкторе класса Book (для соответствия методу EmplaceBack)

Commit:
Реализована функция buildAuthorHistogramFlat

Commit:
Добавлены функции calculateGenreRatings и calculateAverageRating
и испрвлено предупреждение компилятора

Commit:
Добавлены функции sampleRandomBooks и getTopNBy

Commit:
Раскамментирована часть кода в main

Commit:
Реализованы класса-помощники с гетерогенным поиском, добавлен
метод hasAutor

Commit:
Добавлена реализация concept-ов

Commit:
Добавлены модульные тесты, исправлено условие в assert-е, удален один
assert с не корректным условием, внесены изменения в concept

Commit:
Добавлена возможность фильтрации книг по набору предикат с применением
алгоритма all_of

Commit:
Добвлена функция any_of

Commit:
Исправлена ошибка в алгоритме и добавлены модульные тесты

ivan-flink commented on include/book.hpp:
```diff
@@ -11,10 +12,37 @@ enum class Genre { Fiction, NonFiction, SciFi, Biography, Mystery, Unknown };
     return Genre::Unknown;
 }
 
+// Функция не constexpЮ так как может выбрасить исключение
```
Но это не проблема, поскольку тогда просто будет ошибка на этапе компиляции

TarasTsalko replied:
Исправил, сделал функцию StringFromGenre constexpr

ivan-flink commented on include/book.hpp:
```diff
@@ -25,30 +53,28 @@ struct Book {
+          read_count(read_count_) {}
 };
+
+using constBookRef = std::reference_wrapper<const Book>;
```
👍

ivan-flink commented on include/book.hpp:
```diff
@@ -58,5 +84,21 @@ struct formatter<bookdb::Genre, char> {
+        const std::string_view format_template = "title : {}, Author : {}, year : {}, "
+                                                 "genre : {}, rating : {}, read_count : {} ";
+        const std::string genre_str = bookdb::StringFromGenre(b.genre);
+        const std::string book_str = std::vformat(
```
тут будет 2 копии строк, что может быть не слишком эффективно. Лучше сразу писать в format_to

TarasTsalko replied:
Исправил, удалил book_str.
genre_str приходится создать, так как std::make_format_arg "требует"  ссылку не на временный объект, как я понял

ivan-flink commented on include/book_database.hpp:
```diff
@@ -15,12 +19,19 @@ template <BookContainerLike BookContainer = std::vector<Book>>
+    using const_iterator = typename BookContainer::const_iterator;
+    using value_type = typename BookContainer::value_type;
+    using reference = typename BookContainer::reference;
+    using AuthorContainer = std::unordered_set<std::string_view, TransparentStringHash, TransparentStringEqual>;
```
Основная операция - поиск, поэтому хеш таблица - наилучший вариант👍

Но тут есть одна очень хитрая ошибка - сейчас в классе BookDatabase никто не владеет списком авторов. К сожалению, без borrow checker или мощного стат анализатора её не поймать.

Проявление проблемы - код ниже
![image](https://github.com/user-attachments/assets/c3e53e48-eb5e-482c-8170-1078d1418449)

ivan-flink replied:
А решив эту проблему проверь методы добавления элементов

TarasTsalko replied:
Исправил

ivan-flink commented on include/comparators.hpp:
```diff
@@ -1,9 +1,21 @@
+};
+
+struct LessByPopularity {
+    // используем >, так как поплярные книги должны находится в верху контейнера (списка)
```
Я бы сказал, что это тогда greater

TarasTsalko replied:
Исправил, переименовал в GreaterByPopularity и GreaterByRating

ivan-flink commented on include/concepts.hpp:
```diff
@@ -1,25 +1,46 @@
 
 template <typename T>
-concept BookContainerLike = true;
+concept BookContainerLike = requires(T cont) {
```
👍
Хороший концепт

ivan-flink commented on include/concepts.hpp:
```diff
@@ -1,25 +1,46 @@
+// TODO: уточнить у ревьювера где его можно пременить
 template <typename S, typename I>
-concept BookSentinel = true;
+concept BookSentinel = std::sentinel_for<S, I>;
```
Идея конца диапазона - это расширение итераторов. С итераторами можно использовать только объект такого же типа, а Sentinel позволяет задать условие окончания с использованием другого типа. Это используется в диапазонах (4 модуль). Пока в данном коде это можно применить для функции `filterBooks`, чтобы указать, что тип конца диапазона можно задать не только итератором, а ещё и Sentinel

```cpp
template <BookIterator T>
auto filterBooks(T beginIt, auto BookSentinel endIt, auto &f) {
```

TarasTsalko replied:
Добавил BookSentinel к параметру endIt функции filterBooks

ivan-flink commented on include/concepts.hpp:
```diff
@@ -1,25 +1,46 @@
+// TODO: уточнить у ревьювера где его можно пременит
 template <typename P>
-concept BookPredicate = true;
+concept BookPredicate = requires() {
```
Тут проще. Это можно использовать в том же `filterBooks`.

Тут возможно стоило дописать, что это унарный предикат. Тогда можно создать такой концепт

```cpp
concept BookPredicate= requires(P pred) {
    { pred(std::declval<Book>()) } -> std::convertible_to<bool>;
};
```

ivan-flink replied:
И использовать в all_of

TarasTsalko replied:
Добавил concept в функции  all_of и any_of

ivan-flink commented on include/filters.hpp:
```diff
@@ -1,11 +1,51 @@
 
+template <typename... Predicates>
+auto all_of(Predicates... preds) {
+    return [=](const Book &value) { return (preds(value) && ...); };
```
чуть более удачное решение - выполнить захват по ссылке или переместить предикаты внутрь лямбды

TarasTsalko replied:
Исправил, переместил предикаты внутрь лямбды, используя move-семантику

ivan-flink commented on include/filters.hpp:
```diff
@@ -1,11 +1,51 @@
+    return [=](const Book &value) { return (preds(value) || ...); };
+}
+
+inline auto YearBetween(int startPeriudYear, int endPeriudYear) {
```
В целом такие простые функции обычно можно сделать constexpr

inline не обязателен. Это лишь рекомендация компилятору + современные компиляторы уже всё хорошо инлайнят, поэтому надо либо force_inline использовать, либо применять inline только для обхода ODR ограничения

TarasTsalko replied:
Исправил, сделал функции constexpr

ivan-flink commented on include/filters.hpp:
```diff
@@ -1,11 +1,51 @@
+
+template <BookIterator T>
+auto filterBooks(T beginIt, T endIt, auto &f) {
+    std::vector<constBookRef> results;
```
constBookRef - алиас, можно с большой буквы назвать как тип

TarasTsalko replied:
Исправил, переименовал алиас в СonstBookRef

ivan-flink commented on include/filters.hpp:
```diff
@@ -1,11 +1,51 @@
+template <BookIterator T>
+auto filterBooks(T beginIt, T endIt, auto &f) {
+    std::vector<constBookRef> results;
+    results.reserve(std::distance(beginIt, endIt));
```
👍

ivan-flink commented on include/filters.hpp:
```diff
@@ -1,11 +1,51 @@
+auto filterBooks(T beginIt, T endIt, auto &f) {
+    std::vector<constBookRef> results;
+    results.reserve(std::distance(beginIt, endIt));
+    while (beginIt != endIt) {
+        if (f(*beginIt))
+            results.emplace_back(*beginIt);
+        beginIt++;
+    }
```
после изучения 4 модуля можно будет полностью избавиться от сырых циклов и ручной итерации, применив диапазоны

ivan-flink commented on include/heterogeneous_lookup.hpp:
```diff
@@ -1,14 +1,29 @@
 namespace bookdb {
 
-struct TransparentStringLess {};
+struct TransparentStringLess {
```
могу ошибаться, но при использовании прозрачного компаратора через него будут идти все сравнения, поэтому 2 string_view работать не будут

TarasTsalko replied:
Проверил, все три перегрузки используются.
bool operator()(std::string_view lhv, const std::string &rhv) const { return lhv < rhv; }
bool operator()(const std::string &lhv, std::string_view rhv) const { return lhv < rhv; }
bool operator()(const std::string &lhv, const std::string &rhv) const { return lhv < rhv; }
Каждая необходима для компиляции.
Если я правильно понял вопрос.

ivan-flink replied:
Тут имел ввиду, что если у нас вот такой контейнер `std::set<std::string_view, Comparator>`, то при использовании такого компаратора при вызове `container.contains("str");`
будет создан временный объект `std::string`, потому что есть только одна перегрузка, которая принимает слева `std::string_view`, а справа const lvalue ref. Поскольку есть прямое преобразование из `const char[]` в `std::string`, то на входе в функцию будет создан временный rvalue std::string, который связывается с `const lvalue ref`

При добавлении перегрузки `operator()(std::string_view, std::string_view)` временный объект будет не нужен

ivan-flink commented on include/heterogeneous_lookup.hpp:
```diff
@@ -1,14 +1,29 @@
+};
 
-struct TransparentStringHash {};
+struct TransparentStringHash {
+    using is_transparent = void;
+    std::size_t operator()(const std::string &obj) const { return std::hash<std::string>()(obj); };
+    std::size_t operator()(std::string_view obj) const { return std::hash<std::string_view>()(obj); };
```
👍

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,150 @@
 namespace bookdb {
 
+template <typename Comparator>
+using histogramMap = std::flat_map<std::string, size_t, Comparator>;
```
Мало данных и частый поиск - flat контейнеры👍

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,150 @@
+auto buildAuthorHistogramFlat(const BookDatabase<T> &cont, Comparator comp = {}) {
+    histogramMap<Comparator> histogram;
+    auto begIt = cont.cbegin();
+    while (begIt != cont.cend()) {
```
не стоит использовать сырые циклы, если можно итерироваться по контейнеру через range-based-for.

Двойной поиск (find + operator[])

Всё тело цикла можно переписать в 2 строчки. Я предлагаю тебе внимательно ознакомиться с документацией этого метода, чтобы лаконично переписать код

https://en.cppreference.com/w/cpp/container/flat_map/try_emplace.html

TarasTsalko replied:
Переписал с использованием :try_emplace

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,150 @@
+    // нужно будет менять и функция
+    genreRatingMapImpl<Comparator> genreRating;
+    while (begIt != endIt) {
+        std::string genreStr = bookdb::StringFromGenre(begIt->genre);
```
Тут конвертация избыточна. Мы как раз используем enum для быстрой работы с жанром

TarasTsalko replied:
Сделал  ключ в map enum(Genre), убрал избыточную конвертацию

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,150 @@
+        std::string genreStr = bookdb::StringFromGenre(begIt->genre);
+        auto it = genreRating.find(genreStr);
+        if (it == genreRating.end()) {
+            genreRating[std::move(genreStr)] = std::make_pair(begIt->rating, 1);
```
двойной поиск

TarasTsalko replied:
Применил try_emplace, убрал двойной поиск

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,150 @@
+    }
+
+    genreRatingMap<Comparator> genreRatingOut;
+    std::transform(genreRating.begin(), genreRating.end(), std::inserter(genreRatingOut, genreRatingOut.end()),
```
Молодец, что применил transform!

В функциях подсчёта статистики бывает полезно создать внутренний класс для удобства работы, например такой
```cpp
    class GenreStats {
    public:
        void Add(double rating);
        double Average() const;
    private:
        double total_rating = 0.0;
        size_t count = 0;
    };
```

TarasTsalko replied:
Переписал код использую вспомогательный класс GenreStats

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,150 @@
+template <BookContainerLike T>
+auto calculateAverageRating(BookDatabase<T> &cont) {
+
+    return std::accumulate(cont.begin(), cont.end(), 0.0,
```
поскольку порядок не важен - используем reduce

TarasTsalko replied:
Переписал с использованием std::transform_reduce

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,150 @@
+
+template <BookContainerLike T>
+    requires BookIterator<typename T::const_iterator>
+auto sampleRandomBooks(const BookDatabase<T> &cont, size_t N) {
```
Попробуй переписать. используя алгоритм с говорящим названием `std::sample`

TarasTsalko replied:
Переписал с использованием std::sample

ivan-flink replied:
👍

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,150 @@
+    if (cont.size() < N)
+        throw std::runtime_error(std::format(
+            "The number of books {} transferred exceeds the number of books in the database {}\n", N, cont.size()));
+    std::stable_sort(cont.begin(), cont.end(), comp);
```
Хорошая идея, использовать stable_sort, но работает не оптимально для маленьких N.

Как идея для улучшения, попробуй переписать, используя nth_element, например, когда log(cont.size()) >2 * N

TarasTsalko replied:
Реализовал функцию  getTopNBy с применением std::nth_element

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,150 @@
+
+namespace std {
+template <>
+struct formatter<bookdb::histogramMap<bookdb::TransparentStringLess>, char> {
```
👍

ivan-flink commented on src/main.cpp:
```diff
@@ -44,32 +44,46 @@ int main() {
 
+    std::print("\n\nTop 3 random books:\n");
+    auto randomBooks = sampleRandomBooks(db, 3);
+    std::for_each(randomBooks.cbegin(), randomBooks.cend(), [](const auto &v) { std::print("{}\n", v.get()); });
```
Здорово, что используешь cbegin / cend

ivan-flink commented on src/main.cpp:
```diff
@@ -44,32 +44,46 @@ int main() {
+        std::print("\n\nThere are books by {} in the database\n", author);
+
+    {
+        auto comp = all_of(YearBetween(1900, 1999), RatingAbove(4.5));
```
👍

ivan-flink commented on tests/db_test.cpp:
```diff
@@ -1,3 +1,179 @@
+    EXPECT_EQ(ref.genre, Genre::Fiction);
+    EXPECT_EQ(ref.author, "Harper Lee");
+    EXPECT_EQ(ref.year, 1960);
+    EXPECT_DOUBLE_EQ(ref.rating, 4.8);
```
👍

ivan-flink commented on tests/db_test.cpp:
```diff
@@ -1,3 +1,179 @@
-TEST(TestComponentName, SimpleCheck) { EXPECT_EQ(1 + 1, 2); }
+#include "comparators.hpp"
+#include "filters.hpp"
+#include "statsistics.hpp"
```
Хорошие тесты 👍

Commit:
Внесена часть правок по ревью: скорректированы имена компораторов,
свободные функции сделаны constexpr, убрано копирование предекатов, в
лямбда функции, к параметрам функции добавлены концепты

Commit:
Правки по ревью: исправлена ​​ошибка, что имена авторов сохранялись
в unordred_set как string_view и изменялись при изменении исходного объекта

Commit:
добавлена перегрузка для begin и для end, для итерации с помощью range-based for

Commit:
Добавлен компоратор LessByGenre для возможности использовать Genre
в flat_map в качестве ключа

Commit:
Правки по ревью: функция buildAuthorHistogramFlat переписына
с использованием try_emplace, чтобы избежать избыточного
поиска при добавлении элемента;
аналогично для функции calculateGenreRatings;
функция sampleRandomBooks переписына с использованием функции std::sample

Commit:
Правки по ревью: calculateAverageRating переписына с применением
std::transform_reduce, исправленна ошибка в случаи если прищла пустая
база книг, добавлен модульный тест.

Commit:
Функция getTopNBy реализована с применением std::nth_element

Commit:
Убрано создание временного объекта const std::string book_str

Commit:
Реализован класс GenreStats для удобства подсчета статистики в функции
calculateGenreRatings и подкорректирован комментарий в коде

ivan-flink commented on include/book.hpp:
```diff
@@ -8,13 +9,37 @@ namespace bookdb {
     return Genre::Unknown;
 }
 
+constexpr std::string StringFromGenre(Genre g) {
```
👍

ivan-flink replied:
Тут сначала не заметил один нюанс. Поскольку все строки "Fiction", ... будут лежать на стеке в программе, то если ты возвращаешь `std::string`, то ты каждый раз из строк на стеке создаёшь строку в куче и возвращаешь её, а возможно, вызывающий код просто хотел её прочитать. Решение C++17 - возвращать `const std::string_view`

Точно также, как мы теперь можем в классе писать `static const constexpr std::string_view s = "My String";`.
Таким образом, если тебе надо только прочитать строку - всё хорошо, ты создал только лёгкий вью на стеке. Если тебе надо получить настоящий `std::string`, то ты в вызывающем коде просто сконструируешь его из вью

ivan-flink commented on include/book.hpp:
```diff
@@ -58,5 +81,23 @@ struct formatter<bookdb::Genre, char> {
+
+        // genre_str приходится создать, так как std::make_format_arg "требует"  ссылку не на времменый объект, как я понял
+        const std::string genre_str = StringFromGenre(b.genre);
+        return format_to(fc.out(), "{}",
```
Да, видимо временный объект создавать надо.

Тут у меня несколько комментариев. 

1. `std::make_format_args` если я правильно помню, нужен для динамического создания аргументов, но тут ты их можешь сам написать вот так
```cpp
format_to(fc.out(), "Title: {}, Author: {}, Year: {}, Genre: {}, Rating: {:.2f}, Read Count: {}",
                         book.title, book.author, book.year, book.genre, book.rating, book.read_count);
```

2. Основная фишка `formatter` - если ты уже определил `formatter<bookdb::Genre, char>`, то ты можешь его просто использовать внутри `formatter<bookdb::Book, char>`

То есть, в коде выше я просто передаю `book.genre` в `format_to` а он уже сам разберётся, что форматирование этого enum уже написано в `formatter<bookdb::Genre, char>`

ivan-flink commented on include/book_database.hpp:
```diff
@@ -29,7 +41,52 @@ class BookDatabase {
+
     // Ваш код здесь
+private:
+    // функция копируют информацию об авторе и обновляет ссылку в объекте типа Book
+    void CopyAndAdd(Book &ref) {
+        const auto insertRes = authors_.emplace(ref.author);
+        ref.author = *insertRes.first;
+    }
```
👍
Чтобы не писать .first можно сделать structure-binding
```cpp
const auto [it, _ ] - ...;
ref.author = *it;
```

ivan-flink commented on include/book_database.hpp:
```diff
@@ -15,12 +20,19 @@ template <BookContainerLike BookContainer = std::vector<Book>>
+    using const_iterator = typename BookContainer::const_iterator;
+    using value_type = typename BookContainer::value_type;
+    using reference = typename BookContainer::reference;
+    using AuthorContainer = std::unordered_set<std::string, TransparentStringHash, TransparentStringEqual>;
```
Здесь мы как раз применили паттерн Lightweight - тяжёлые строки хранятся в единственном экземпляре, а в книгах только лёгкие `std::string_view`

ivan-flink commented on include/filters.hpp:
```diff
@@ -1,11 +1,57 @@
+    };
+}
+
+template <BookPredicate... Predicates>
```
👍

ivan-flink commented on include/filters.hpp:
```diff
@@ -1,11 +1,57 @@
+template <BookPredicate... Predicates>
+auto any_of(Predicates... preds) {
+    auto preds_tuple = std::make_tuple(std::forward<Predicates>(preds)...);
+    return [p = std::move(preds_tuple)](const Book &value) {
+        return std::apply([&](auto &&...pred) { return (pred(value) || ...); }, p);
+    };
```
👍 Так тоже можно, хотя мне кажется твой изначальный вариант был лучше, просто в нём можно предикаты  переместить

`return [... ps = std::move(preds)](const Book &book) { return ( ps(book) || ... ); };`

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,151 @@
+auto buildAuthorHistogramFlat(const BookDatabase<T> &cont, Comparator comp = {}) {
+    histogramMap<Comparator> histogram;
+    for (const auto &val : cont) {
+        auto [it, inserted] = histogram.try_emplace(std::string(val.author), 1);
+        if (!inserted)
+            it->second++;
```
Можно без `if`

```cpp
        auto [it, _] = hist.emplace(book.author, 0u);
        ++it->second;
```

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,151 @@
+
+    using genreRatingMapImpl = std::flat_map<bookdb::Genre, GenreStats, Comparator>;
+    genreRatingMapImpl genreRating;
+    while (begIt != endIt) {
```
если есть пара итераторов, то можно использовать `std::for_each`

Вот один из вариантов 
```cpp
    std::for_each(first, last, [&stats](const auto& v) {
        auto [it, _] = stats.try_emplace(v.genre);
        it->second.Add(v.rating);
    });
```

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,151 @@
+template <BookContainerLike T>
+auto calculateAverageRating(BookDatabase<T> &cont) {
+
+    const double sum = std::transform_reduce(cont.begin(), cont.end(), 0.0, std::plus<>(),
```
👍

ivan-flink commented on include/statsistics.hpp:
```diff
@@ -1,18 +1,151 @@
+            "The number of books {} transferred exceeds the number of books in the database {}\n", N, cont.size()));
+
+    // если количество книг, которые нужно "извлечь" сильно меньше чем книг в базе
+    if (std::log(size) > 2 * N) {
+        std::nth_element(cont.begin(), cont.begin() + N, cont.end(), comp);
+        auto topN = std::vector<ConstBookRef>(cont.begin(), cont.begin() + N);
+        std::stable_sort(topN.begin(), topN.end(), comp);
+        return topN;
+    } else {
+        std::stable_sort(cont.begin(), cont.end(), comp);
+        return std::vector<ConstBookRef>(cont.begin(), cont.begin() + N);
+    }
```
Отличная реализация!
